### PR TITLE
Add Coefficient*[]

### PR DESCRIPTION
**Pull request by hailongnk**
I am planning to add Coefficient[] and CoefficientList[]
This work in in-progress, this pull request is created to trigger automated tests

**Notes for Coefficient:**
- Modulus is not supported
- 2 test cases are failed because of Sympy bug
<pre>Coefficient[(x + 2)/(y - 3) + (x + 3)/(y - 2), z, 0]
The correct result should be (2 + x) / (-3 + y) + (3 + x) / (-2 + y)
but Sympy 1.0 returns 0</pre>
<pre>Coefficient[a x^2 + b y^3 + c x + d y + 5, x, 0]
Correct result: 5 + b y ^ 3 + d y
Sympy 1.0 returns: 5</pre>

This bug has been fixed in current [sympy-master](https://github.com/sympy/sympy)

### CODE REVIEW

**Comment by sn6uv** [comment]
As a note, if you register with TravisCI you can get the CI builds on your own branch without triggering a PR. For example: https://travis-ci.org/sn6uv/Mathics. That said, I have no problem with incomplete PRs just make sure to ping me (slack is best) if you're done and I don't get to it.

**Comment by hailongnk** [comment]
Thanks for the hint :)

**Comment by hailongnk** [comment]
2 first form of CoefficientList have been implemented

File: mathics/builtin/algebra.py [hunk 1]
```diff
@@ -789,4 +789,260 @@ def apply(self, expr, v, evaluation):
...
+                    if len(dims) == 0:
+                        return from_sympy(poly.nth(*[i for i in exponents]))
+                    
+                    result = Expression('List')
+                    first_dim = dims[0]
+                    for i in range(first_dim+1):
+                        exponents.append(i)
+                        subs = _nth(poly, dims[1:], exponents)
+                        result.leaves.append(subs)
+                        if len(exponents): exponents.pop()
```
Comment (sn6uv) [inline mathics/builtin/algebra.py]: do you need the if check here?

File: mathics/builtin/algebra.py [hunk 2]
```diff
@@ -789,4 +789,260 @@ def apply(self, expr, v, evaluation):
...
+                        *[_coefficient(self.__class__.__name__,expr, form, Integer(n), evaluation)
+                          for n in range(dimensions[0]+1)])
+            elif form.has_form('List', 1):
+                form = form.leaves[0]
+                return Expression('List',
+                        *[_coefficient(self.__class__.__name__, expr, form, Integer(n), evaluation)
+                          for n in range(dimensions[0]+1)])
+            else:
+                def _nth(poly, dims, exponents):
+                    if len(dims) == 0:
```
Comment (sn6uv) [inline mathics/builtin/algebra.py]: alternatively, `if not dims:`

File: mathics/builtin/algebra.py [hunk 3]
```diff
@@ -789,4 +789,260 @@ def apply(self, expr, v, evaluation):
...
+    
+    def apply_noform(self, expr, evaluation):
+        'CoefficientList[expr_]'
+        return evaluation.message('CoefficientList', 'argtu')
+    
+    def apply(self, expr, form, evaluation):
+        'CoefficientList[expr_, form_]'
+        vars = [form] if not form.has_form('List', None) else [v for v in form.leaves]
+        
+        # form is not a variable
```
Comment (sn6uv) [inline mathics/builtin/algebra.py]: why?

File: mathics/builtin/algebra.py [hunk 4]
```diff
@@ -789,4 +789,260 @@ def apply(self, expr, v, evaluation):
...
+            return evaluation.message('CoefficientList', 'poly', expr)
+        
+        try:
+            sympy_poly, sympy_opt = sympy.poly_from_expr(sympy_expr, sympy_vars)
+            dimensions = [sympy_poly.degree(x) if x in sympy_poly.gens else 0 for x in sympy_vars]
+            
+            # single & multiple variables cases
+            if not form.has_form('List', None):
+                return Expression('List', 
+                        *[_coefficient(self.__class__.__name__,expr, form, Integer(n), evaluation)
```
Comment (sn6uv) [inline mathics/builtin/algebra.py]: indented with 8 spaces, should be 4

File: mathics/builtin/algebra.py [hunk 5]
```diff
@@ -789,4 +789,260 @@ def apply(self, expr, v, evaluation):
...
+                          for n in range(dimensions[0]+1)])
+            elif form.has_form('List', 1):
+                form = form.leaves[0]
+                return Expression('List',
+                        *[_coefficient(self.__class__.__name__, expr, form, Integer(n), evaluation)
```
Comment (sn6uv) [inline mathics/builtin/algebra.py]: indentation.

File: mathics/builtin/algebra.py (hunk 3 above)
Comment (hailongnk) [inline mathics/builtin/algebra.py]: correct

File: mathics/builtin/algebra.py (hunk 4 above)
Comment (hailongnk) [inline mathics/builtin/algebra.py]: oops, it should be better if I change my tabs setting to 4 spaces as default

File: mathics/builtin/algebra.py (hunk 1 above)
Comment (hailongnk) [inline mathics/builtin/algebra.py]: yes, no need for the if check

**Comment by hailongnk** [comment]
A commit has been submitted to cover your feedback. Thanks

**Pull request merged**